### DISCUSSION TITLE
no_std build with pest_derive fails

### DISCUSSION THREAD

**Discussion by jzbor** [pest-parser/pest#1057] [Q&A] [answered]
**Describe the bug**
I get the following error, when trying to build with no_std:
```
   Compiling lash v0.3.1 (/home/jzbor/Programming/Rust/lash)
error: cannot find macro `format` in this scope
  --> src/typing.rs:48:13
   |
48 |             format!("{}", ('A' as usize + n) as u8 as char)
   |             ^^^^^^

error: cannot find macro `format` in this scope
  --> src/typing.rs:50:13
   |
50 |             format!("T{}", n)
   |             ^^^^^^

error: cannot find macro `format` in this scope
   --> src/typing.rs:154:58
    |
154 |             Variable(x) => Ok(self.inst(env.get(x).ok_or(format!("unable to instantiate {}", x))?.clone())),
    |                                                          ^^^^^^

error: cannot find macro `format` in this scope
   --> src/typing.rs:126:13
    |
126 |         Err(format!("unable to unify {} with {}", t1, t2))
    |             ^^^^^^

error: cannot find macro `format` in this scope
  --> src/typing.rs:77:21
   |
77 |                 Err(format!("occurs check failed - {} versus {}", tv.borrow(), tv2.borrow()))
   |                     ^^^^^^

error: cannot find macro `format` in this scope
  --> src/typing.rs:65:21
   |
65 |                 Err(format!("occurs check failed - {} == {}", tv.borrow(), tv2.borrow()))
   |                     ^^^^^^

error[E0433]: failed to resolve: could not find `std` in the list of imported crates
  --> src/parsing.rs:17:10
   |
17 | #[derive(Parser)]
   |          ^^^^^^ could not find `std` in the list of imported crates
   |
   = note: this error originates in the derive macro `Parser` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `std` in the list of imported crates
  --> src/parsing.rs:17:10
   |
17 | #[derive(Parser)]
   |          ^^^^^^ could not find `std` in the list of imported crates
   |
   = note: this error originates in the derive macro `Parser` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this module
   |
3  + use core::result;
   |

For more information about this error, try `rustc --explain E0433`.
error: could not compile `lash` (lib) due to 8 previous errors
```

**To Reproduce**
Clone https://github.com/jzbor/lash/tree/pest
Run `cargo build --lib --no-default-features`

**Expected behavior**
From what I understand this is a supported setup, so it should just build and use `core` and `alloc` where necessary.

**Comment by tomtau** [comment]
For pest and pest_derive:  https://github.com/jzbor/lash/blob/pest/Cargo.toml#L23

they should be specified with

```
pest = { version = "2.7.14", default-features = false }
pest_derive = { version = "2.7.14", default-features = false }
```

I believe when one calls `cargo build --lib --no-default-features`, it only applies that `--no-default-features` to the crate that's being built, not recursively to its dependencies. In the features, it should also specify this:

```
[features]
default = ["std"]
std = [ "rustyline", "clap", "colored", "humantime", "pest/std", "pest_generator/std" ]
```

so that the std feature is on by default for pest dependencies in that crate.

**Reply by jzbor** [reply]
Thanks a lot, I forgot about that...

**Reply by jzbor** [reply]
Hm, now I get "could not find 'alloc' in the list of imported crates" on the  `#[derive(Parser)]`

**Reply by jzbor** [reply]
Maybe I have to build alloc in the crate, I'll see...

**Reply by jzbor** [reply]
Hm I tried with my OS crate, which builds alloc, but I get the same issue

**Reply by tomtau** [reply]
Maybe cargo needs some extra flags to be passed? e.g.
```
 cargo build -j1 -Z build-std=core,alloc --lib --no-default-features
```

instead of just `cargo build --lib --no-default-features`

**Reply by jzbor** [reply]
Hey, sorry for taking so long. I tried to run that command on the respective branch and it failed with:
```
error[E0433]: failed to resolve: could not find `alloc` in the list of imported crates
  --> src/parsing.rs:16:10
   |
16 | #[derive(Parser)]
   |          ^^^^^^ could not find `alloc` in the list of imported crates
   |
   = note: this error originates in the derive macro `Parser` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `lash` (lib) due to 1 previous error
```

You can checkout that branch [here](https://github.com/jzbor/lash/tree/pest)

**Reply by jzbor** [reply]
I believe [this](https://stackoverflow.com/questions/71675411/refer-to-an-extern-crate-in-macro-expansion) StackOverflow-Thread may be related

**Comment by jzbor** [comment]
@tomtau are there any updates on this? I do believe it is an actual issue with pest, so it would be great if you could either provide a solution or convert this back to an issue so it can be tracked and found by others.

**Reply by tomtau** [reply]
I wasn't able to reproduce it before (it build ok with `cargo build -j1 -Z build-std=core,alloc --lib --no-default-features`); I'll try again in a fresh environment

**Reply by tomtau** [reply, marked as answer]
@jzbor 

1. I checked out the branch you posted: https://github.com/jzbor/lash/tree/pest
2. I added `extern crate alloc` to `src/lib.rs`
```
diff --git a/src/lib.rs b/src/lib.rs
index 605d50c..ae0daa5 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,5 +1,5 @@
 #![cfg_attr(not(feature = "std"), no_std)]
-
+extern crate alloc;
 pub mod debruijn;
 pub mod environment;
 pub mod error;
```

3. it builds fine:
```
$ cargo build -j1 -Z build-std=core,alloc --lib --no-default-features
   Compiling lash v0.3.1 (/workspaces/lash)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.22s
```

**Reply by jzbor** [reply]
Oh, I am very sorry, if I have wasted your time. I was not aware that `extern crate alloc;` was also required at the crate root.

**Reply by stefnotch** [reply]
I ran into the same issue, and adding `extern crate alloc;` fixed it.
Would be appreciated if this were documented somewhere next to the feature flag.
